### PR TITLE
Simulate port is already allocated (#62)

### PR DESCRIPTION
**Pull request by luanqi521-s**
Signed-off-by: Luanqi <949807469@qq.com>

### CODE REVIEW

File: pkg/server/chaosd/network.go [hunk 1]
```diff
@@ -296,3 +300,57 @@ func (s *Server) recoverDNSServer(attack *core.NetworkCommand) error {
...
 }
+
+
+func (s *Server) applyPortOccupied(attack *core.NetworkCommand) error {
+
+	if len(attack.Port) == 0 {
+		return nil
+	}
+
+	//todo:修改对已启用端口检测
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:311]: it's better to use English as comment

File: pkg/server/chaosd/network.go [hunk 2]
```diff
@@ -296,3 +300,57 @@ func (s *Server) recoverDNSServer(attack *core.NetworkCommand) error {
...
+   err, flag := checkPortIsListened(attack.Port)
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:312]: You can run `make fmt` to format the code

File: pkg/server/chaosd/network.go [hunk 3]
```diff
@@ -296,3 +300,57 @@ func (s *Server) recoverDNSServer(attack *core.NetworkCommand) error {
...
+    if err == nil && flag == false  {
+		c := fmt.Sprintf("nc -l %s", attack.Port)
+		cmd := exec.Command("sh", "-c", c)
+
+		err = cmd.Start()
+		if err != nil {
+			errors.WithStack(err)
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:319]:
```suggestion
			return errors.WithStack(err)
```

File: pkg/server/chaosd/network.go [hunk 4]
```diff
@@ -296,3 +300,57 @@ func (s *Server) recoverDNSServer(attack *core.NetworkCommand) error {
...
+		}
+		fmt.Println("exec cmd success")
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:321]: how about using `log.Info`, just like other logs

File: pkg/server/chaosd/network.go [hunk 5]
```diff
@@ -296,3 +300,57 @@ func (s *Server) recoverDNSServer(attack *core.NetworkCommand) error {
...
+
+}
+
+func checkPortIsListened(port string) (error, bool) {
+	checkStatement := fmt.Sprintf("array=lsof -i:%s | awk '{print $2}' | grep -v PID;echo ${array[@]}", port)
+	cmd := exec.Command("sh","-c", checkStatement)
+
+	output, err1 := cmd.Output()
+	if err1 != nil {
+		fmt.Println(err1)
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:335]: ditto, can use `log.Error`

File: pkg/server/chaosd/network.go [hunk 6]
```diff
@@ -296,3 +300,57 @@ func (s *Server) recoverDNSServer(attack *core.NetworkCommand) error {
...
+		return err1, false
+	}
+	if len(output) != 0 {
+		return nil, true
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:339]: can print `output` in the log for debug

Comment (WangXiangUSTC) [reply pkg/server/chaosd/network.go:339]: and don't need to return

File: pkg/server/chaosd/network.go [hunk 7]
```diff
@@ -309,19 +309,27 @@ func (s *Server) applyPortOccupied(attack *core.NetworkCommand) error {
...
-   err, flag := checkPortIsListened(attack.Port)
-    if err == nil && flag == false  {
+   //err, flag := checkPortIsListened(attack.Port)
+
+	checkStatement := fmt.Sprintf("array=lsof -i:%s | awk '{print $2}' | grep -v PID;echo ${array[@]}", attack.Port)
+	cmd := exec.Command("sh","-c", checkStatement)
+
+	output, err := cmd.Output()
+	fmt.Println(string(output))
+	return errors.WithStack(err)
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:319]: should return err when err is not nil

File: pkg/server/chaosd/network.go [hunk 8]
```diff
@@ -296,3 +300,57 @@ func (s *Server) recoverDNSServer(attack *core.NetworkCommand) error {
...
+		return nil
+	} else {
+		return err
+	}
+
+}
+
+func checkPortIsListened(port string) (error, bool) {
+	checkStatement := fmt.Sprintf("array=lsof -i:%s | awk '{print $2}' | grep -v PID;echo ${array[@]}", port)
+	cmd := exec.Command("sh","-c", checkStatement)
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/server/chaosd/network.go:331]: @muse-dev ignore

File: pkg/server/chaosd/network.go [hunk 9]
```diff
@@ -310,47 +312,53 @@ func (s *Server) applyPortOccupied(attack *core.NetworkCommand) error {
 		return nil
 	}
 
-    err, flag := checkPortIsListened(attack.Port)
-    if err == nil && flag == true  {
+    flag, err := checkPortIsListened(attack.Port)
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:315]: do we need to print an error log when err is not nil?

File: pkg/server/chaosd/network.go [hunk 10]
```diff
@@ -310,47 +312,53 @@ func (s *Server) applyPortOccupied(attack *core.NetworkCommand) error {
...
+    if flag && err == nil{
     	return errors.Errorf("port %s has been occupied", attack.Port)
 	}
 	
 	c := fmt.Sprintf("nc -l %s", attack.Port)
-	cmd1 := exec.Command("sh", "-c", c)
-	err = cmd1.Start()
+	cmd := exec.Command("sh", "-c", c)
+	err = cmd.Start()
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:322]: how about using `output, err := cmd.CombinedOutput()`? so we can print output to know the detail when cmd run failed

File: pkg/server/chaosd/network.go [hunk 11]
```diff
@@ -310,47 +312,53 @@ func (s *Server) applyPortOccupied(attack *core.NetworkCommand) error {
...
 	cmd := exec.Command("sh","-c", checkStatement)
 
     stdout, err  := cmd.StdoutPipe()
 	if err != nil {
-		fmt.Println("cmd stdoutpipe get err")
-		return err, false
+		log.Error("checkPortIsListened get cmd stdoutpipe error", zap.Error(err))
+		return false, err
+	}
+	err = cmd.Start()
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:338]: ditto, maybe it is simpler to use `cmd.CombinedOutput`

File: pkg/server/chaosd/network.go [hunk 12]
```diff
@@ -296,3 +303,67 @@ func (s *Server) recoverDNSServer(attack *core.NetworkCommand) error {
...
+
+func (s *Server) applyPortOccupied(attack *core.NetworkCommand) error {
+
+	if len(attack.Port) == 0 {
+		return nil
+	}
+
+	flag, err := checkPortIsListened(attack.Port)
+	if err != nil {
+		if flag {
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:315]: seems that it not necessary to judge the flag here

File: pkg/server/chaosd/network.go [hunk 13]
```diff
@@ -296,3 +303,67 @@ func (s *Server) recoverDNSServer(attack *core.NetworkCommand) error {
...
+
+func (s *Server) recoverPortOccupied(attack *core.NetworkCommand, uid string) error {
+
+	c := fmt.Sprintf("lsof -i:%s | awk '{print $2}' | grep -v PID | xargs kill -9", attack.Port)
+
+	cmd := exec.Command("sh", "-c", c)
+
+	stdout, err := cmd.CombinedOutput()
+	if err != nil {
+		log.Error(cmd.String()+string(stdout), zap.Error(err))
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:364]: how about adding some description for the message, for example:
```suggestion
		log.Error(fmt.Sprintf("execute command: %s, output: %s", cmd.String(), string(stdout)), zap.Error(err))
```

**Review by WangXiangUSTC — COMMENTED** [review]
rest LGTM

**Comment by Andrewmatilde** [comment]
Suggest, [#1](https://github.com/luanqi521-s/chaosd-1/pull/1)

File: pkg/server/chaosd/network.go [hunk 14]
```diff
@@ -357,7 +360,8 @@ func checkPortIsListened(port string) (bool, error) {
 
 func (s *Server) recoverPortOccupied(attack *core.NetworkCommand, uid string) error {
 
-	c := fmt.Sprintf("lsof -i:%s | awk '{print $2}' | grep -v PID | xargs kill -9", attack.Port)
+	//c := fmt.Sprintf("lsof -i:%s | awk '{print $2}' | grep -v PID | xargs kill -9", attack.Port)
+	c := fmt.Sprintf("kill -9 %d", attack.PortPid)
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:364]: https://github.com/chaos-mesh/chaosd/blob/a8213ea516a4ee33d12bd16b92f71eb292a71932/pkg/server/chaosd/stress.go#L102

it is better to add a judgement before kill it, make sure it is the process of httpStartTool

File: tools/main/httpStartTool.go [hunk 15]
```diff
@@ -0,0 +1,45 @@
+package main
```
Comment (WangXiangUSTC) [inline tools/main/httpStartTool.go:1]: do we need to change this tool's name? like `PortOccupyTool`?

**Comment by WangXiangUSTC** [comment]
`tools/main/httpStartTool` this file should be removed

File: tools/main/PortOccupyTool.go [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+	return nil
+}
+
+func generateHttpPort(port string) string {
+	s := fmt.Sprintf(":%s", port)
+	return s
+}
+
+func startHttp(porttoOccupy string) {
```
Comment (WangXiangUSTC) [inline tools/main/PortOccupyTool.go:34]:
```suggestion
func startHttp(portToOccupy string) {
```

File: pkg/server/chaosd/network.go [hunk 17]
```diff
@@ -296,3 +305,81 @@ func (s *Server) recoverDNSServer(attack *core.NetworkCommand) error {
...
+		}
+		return errors.WithStack(err)
+	}
+
+	if flag {
+		return errors.Errorf("port %s has been occupied", attack.Port)
+	}
+
+	args := fmt.Sprintf("-p %s", attack.Port)
+	cmd := bpm.DefaultProcessBuilder("../../../tools/main/PortOccupyTool", args).Build()
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:328]:
```suggestion
	cmd := bpm.DefaultProcessBuilder("PortOccupyTool", args).Build()
```

File: tools/main/PortOccupyTool.go [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"fmt"
+	"github.com/pingcap/errors"
+	"github.com/spf13/cobra"
+	"log"
+	"net/http"
+)
+
+func PortOccupyTool(port string) error {
```
Comment (WangXiangUSTC) [inline tools/main/PortOccupyTool.go:11]: seems that `port` can be deleted

File: tools/main/PortOccupyTool.go [hunk 19]
```diff
@@ -0,0 +1,49 @@
+package main
```
Comment (WangXiangUSTC) [inline tools/main/PortOccupyTool.go:1]: can move this file to `tools/PortOccupyTool.go`

**Comment by WangXiangUSTC** [comment]
LGTM

**Comment by WangXiangUSTC** [comment]
@Andrewmatilde PTAL again

**Review by WangXiangUSTC — APPROVED** [review]

**Review by Andrewmatilde — APPROVED** [review]
LGTM

**Pull request merged**